client/src/modules/blockEditor/actions.ts

```diff
@@ -56,17 +59,77 @@ export const changePointerReference = ({ id, reference }) => {
   };
 };
 
+const onlyOneNodeThatIsPointerExport = nodes => {
+  return (
+    nodes.length === 1
+    &&
+    nodes[0].type === "pointerExport"
+  );
+};
+
+const twoNodesFirstExportSecondEmptyString = nodes => {
+  return (
+    nodes.length === 2
+    &&
+    nodes[0].type === "pointerExport"
+    &&
+    nodes[1].object === "text"
+    && nodes[1].leaves[0].text === ""
+  );
+};
+
 export const exportSelection = () => {
   return async (dispatch, getState) => {
     const { blocks, blockEditor } = await getState();
     const { hoveredItem } = blockEditor;
 
     const block = blocks.blocks.find(b => b.id === hoveredItem.blockId);
+
     const uuid = uuidv1();
     if (block) {
-      const change = block.value
-        .change()
-        .wrapInline({ type: "pointerExport", data: { pointerId: uuid } });
+
+      const value = block.value;
+
+      const initialFragment = value.fragment;
+      const initialTopLevelNodes = initialFragment.nodes.toJSON()[0].nodes;
+
+      const isNestedInPointerExport =
+        onlyOneNodeThatIsPointerExport(initialTopLevelNodes)
+        ||
+        twoNodesFirstExportSecondEmptyString(initialTopLevelNodes);
+
+      const change = value.change();
+
+      // if we are exporting a nested pointer, guarantee that the selection is
+      // not extending to an edge, because selections on the edge of an inline
+      // pick up what comes after
+      if (isNestedInPointerExport) {
+        moveSelectionAwayFromPointerEdge(change);
```

Comment:
Does this function mutate change? If so, is there any way you could write it into an immutable function, that returns a different change object instead?